DM-50841: Butler server authorization using Gafaelfawr groups

dhirving opened the pull request:
Butler server can now authorize access to specific repositories based on the user's Gafaelfawr group memberships.  This will be used to restrict access to Data Preview 1 ahead of its public release.

Prior to this PR, there was no concept of configuration specific to Butler server -- everything was shared with the `DirectButler` configuration.  `pydantic-settings` is now used to configure the server, adding configuration variables for group membership restrictions and the Gafaelfawr endpoint.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Add a minimal Gafaelfawr REST client

This is able to retrieve the list of groups associated with a given user.

dhirving committed:
Add class for authorizing access to repos

Added a GafaelfawrGroupAuthorizer class that checks whether a given user is a member of the groups authorized to access a given repository

dhirving committed:
Add per-repository config for Butler server

Instead of using the same DAF_BUTLER_REPOSITORIES environment variable as DirectButler, add a separate DAF_BUTLER_SERVER_REPOSITORIES environment variable for Butler server with a more complex format.  This will allow us to add more server-only, repository-specific configuration items.

dhirving committed:
Add Gafaelfawr group authorization to server

Butler server now checks if a user belongs to an authorized group before granting them access to a repository

dhirving committed:
Fix test failure caused by stale cached server config

dhirving committed:
Update Docker integration test for new config format

dhirving committed:
Work around pydantic-settings bug

Fix an issue where pydantic-settings would not JSON-decode a nested model class that was declared after the main settings model.

dhirving committed:
Split group authorization to separate dependency

This will make unit testing easier.

dhirving committed:
Split authorization dependency even more

Split out the creation of the cached authorization client object from the dependency that checks group membership.  This will let us more easily inject a mocked authorization dependency for testing.

dhirving committed:
Add status_code property to ButlerServerError

dhirving committed:
Add test for authorization plumbing

dhirving committed:
Add tests for authorization logic

dhirving committed:
Add missing docstrings

dhirving committed:
Remove unused class

dhirving committed:
Increase timeout and add retries for Gafaelfawr

dhirving committed:
Remove authorization check from Butler config endpoint

These endpoints are used by higher-level Butler initialization code that has no knowledge of HTTP authentication or Gafaelfawr, so they need to be accessible without requiring authentication headers.

dhirving committed:
Skip server tests in pipeline builds

rra commented on python/lsst/daf/butler/remote_butler/server/_gafaelfawr.py:
```diff
@@ -0,0 +1,127 @@
+
+    async def get_groups(self, user_token: str) -> list[str]:
+        response = await self._client.get(
+            "/auth/api/v1/user-info", headers=get_authentication_headers(user_token)
```
It's probably slightly more future-proof to move `/auth` here into the base URL. I think the `/auth` prefix will only be used when everything is running in the same domain. The long-term intent is to have Gafaelfawr run under auth.data.lsst.cloud (for example), and in that configuration we'll probably drop the `/auth` prefix.

rra commented on python/lsst/daf/butler/remote_butler/server/_gafaelfawr.py:
```diff
@@ -0,0 +1,127 @@
+        return [group.name for group in info.groups]
+
+
+class _GafaelfawrUserInfo(pydantic.BaseModel):
+    username: str
+    groups: list[_GafaelfawrGroup] | None = None
+
+
+class _GafaelfawrGroup(pydantic.BaseModel):
+    name: str
```
Apologies for having to duplicate these models; will definitely fix this once I have a moment to breathe and make a proper Gafaelfawr client library.

rra reviewed (approved):
Gafaelfawr parts look good with one minor comment.

dhirving committed:
Change definition of Gafaelfawr base URL

Per Russ, the base URL includes the `/auth` part of the path because that may move to a subdomain later.
